runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -7526,15 +7511,7 @@ done:;
 			J9Class *arrayClass = resolvedClass->arrayClass;
 			if (J9_EXPECTED(NULL != arrayClass)) {
 				j9object_t instance = NULL;
-#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
-				if (J9_IS_J9CLASS_FLATTENED(arrayClass)) {
-					instance = _objectAllocate.inlineAllocateIndexableValueTypeObject(_currentThread, arrayClass, (U_32)size);
-				} else
-#endif /* J9VM_OPT_VALHALLA_VALUE_TYPES */
-				{
-					instance = _objectAllocate.inlineAllocateIndexableObject(_currentThread, arrayClass, (U_32)size);
-				}
-
+				instance = VM_VMHelpers::arrayAllocation(_currentThread, &_objectAllocate, arrayClass, (U_32) size);
```

Comment:
you need to add the J9ClassContainsUnflattenedFlattenables check to anewarray


and also in any other place that is allocating an array without the helper where it is unknown what the element type is

Reply:
places like `inlInternalsPrimitiveClone` and `inlArrayNewArrayImpl`

Reply:
actually not `inlInternalsPrimitiveClone`

Reply:
@tajila Do I also omit the check in `anewarray` ? I do unflattenedflattenable check in `J9AllocateIndexableObject`, and in  `anewarray`, `J9AllocateIndexableObject` is called when `NULL == instance`. Will this be enough?

Reply:
Similarly, I also see `inlArrayNewArrayImpl` is calling `allocateIndexableObject`, which is using `J9AllocateIndexableObject` when `inlineArrayAllocation` returns NULL. Should I still do the check in `inlArrayNewArrayImpl` or just let it fall to where `J9AllocateIndexableObject` is called?

Reply:
you need the `unflattenedflattenable` check before using the inlineAllocateArray unless you are certain that the class is not a value type

Reply:
@tajila Now I wrap the `inlineArrayAllocation` with the `unflattenable` check except in `inlInternalsPrimitiveClone`.

Reply:
@tajila Now I wrap the `inlineArrayAllocation` with the `unflattenable` check except in `inlInternalsPrimitiveClone`.